src/query/storages/share/src/share.rs

```diff
@@ -0,0 +1,193 @@
+// Copyright 2022 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::BTreeMap;
+
+use common_exception::Result;
+use common_meta_app::share::ShareDatabaseSpec;
+use common_meta_app::share::ShareSpec;
+use common_meta_app::share::ShareTableSpec;
+use opendal::Operator;
+
+pub const SHARE_CONFIG_PREFIX: &str = "_share_config";
+
+#[derive(serde::Serialize, serde::Deserialize, Clone, Debug, Default, Eq, PartialEq)]
+pub struct ShareSpecVec {
+    share_specs: BTreeMap<u64, ext::ShareSpecExt>,
+}
+
+pub async fn save_share_spec(
```

Comment:
there might have one risk on the write-after-read approach from a s3 file:

not all object storages guarantee read-your-write consistency, if two GRANT SHARE operations on one instance closely, it might risks lost-update.

```
Unlike other cloud providers, Amazon S3 delivers strong read-after-write consistency for any storage request, without changes to performance or availability, without sacrificing regional isolation for applications, and at no additional cost.
```

I guess there're ways to work around this:

1. dump all the tenant's shares data from meta on each save
2. dump each shares data with a standalone file name with version and save the latest version in metastore